File: packages/flutter-im-ui/test/conversation_action_sheet_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:flare_im_ui/src/components/flare_conversation_action_sheet.dart';
import 'package:flare_im_ui/src/tokens/flare_strings.dart';

void main() {
  const base = FlareConversationActionSnapshot(id: 'c1', title: '设计评审');
  const all = FlareConversationActionCapabilities(
    pin: true,
    mute: true,
    markRead: true,
    archive: true,
    hide: true,
    delete: true,
  );
  const every = FlareConversationActionCapabilities(
    pin: true,
    mute: true,
    markRead: true,
    markUnread: true,
    archive: true,
    hide: true,
    clearHistory: true,
    delete: true,
  );
  List<FlareConversationAction> ids(
    FlareConversationActionSnapshot c,
    FlareConversationActionCapabilities caps,
  ) => conversationActions(c, caps).map((e) => e.action).toList();

  group('conversationActions', () {
    test('nothing without capabilities', () {
      expect(ids(base, const FlareConversationActionCapabilities()), isEmpty);
    });
    test('each capability reveals exactly its action', () {
      expect(ids(base, const FlareConversationActionCapabilities(pin: true)), [
        FlareConversationAction.pin,
      ]);
      expect(ids(base, const FlareConversationActionCapabilities(mute: true)), [
        FlareConversationAction.mute,
      ]);
      expect(
        ids(base, const FlareConversationActionCapabilities(archive: true)),
        [FlareConversationAction.archive],
      );
      expect(ids(base, const FlareConversationActionCapabilities(hide: true)), [
        FlareConversationAction.hide,
      ]);
      expect(
        ids(
          base,
          const FlareConversationActionCapabilities(clearHistory: true),
        ),
        [FlareConversationAction.clearHistory],
      );
      expect(
        ids(base, const FlareConversationActionCapabilities(delete: true)),
        [FlareConversationAction.delete],
      );
    });
    test('inverts pin/mute/archive by state', () {
      expect(
        ids(
          const FlareConversationActionSnapshot(
            id: 'c',
            title: 't',
            pinned: true,
          ),
          const FlareConversationActionCapabilities(pin: true),
        ),
        [FlareConversationAction.unpin],
      );
      expect(
        ids(
          const FlareConversationActionSnapshot(
            id: 'c',
            title: 't',
            muted: true,
          ),
          const FlareConversationActionCapabilities(mute: true),
        ),
        [FlareConversationAction.unmute],
      );
      expect(
        ids(
          const FlareConversationActionSnapshot(
            id: 'c',
            title: 't',
            archived: true,
          ),
          const FlareConversationActionCapabilities(archive: true),
        ),
        [FlareConversationAction.unarchive],
      );
    });
    test('markRead only when unread > 0', () {
      expect(
        ids(base, const FlareConversationActionCapabilities(markRead: true)),
        isEmpty,
      );
      expect(
        ids(
          const FlareConversationActionSnapshot(
            id: 'c',
            title: 't',
            unreadCount: 3,
          ),
          const FlareConversationActionCapabilities(markRead: true),
        ),
        [FlareConversationAction.markRead],
      );
    });
    test('offers markUnread only when nothing is unread', () {
      const unread = FlareConversationActionSnapshot(
        id: 'c',
        title: 't',
        unreadCount: 3,
      );
      const both = FlareConversationActionCapabilities(
        markRead: true,
        markUnread: true,
      );
      expect(ids(base, both), [FlareConversationAction.markUnread]);
      expect(ids(unread, both), [FlareConversationAction.markRead]);
      expect(
        ids(
          unread,
          const FlareConversationActionCapabilities(markUnread: true),
        ),
        isEmpty,
      );
    });
    test('stable order, clearHistory then delete last and danger', () {
      final entries = conversationActions(
        const FlareConversationActionSnapshot(
          id: 'c',
          title: 't',
          unreadCount: 2,
        ),
        const FlareConversationActionCapabilities(
          pin: true,
          mute: true,
          markRead: true,
          archive: true,
          hide: true,
          clearHistory: true,
          delete: true,
        ),
      );
      expect(entries.map((e) => e.action).toList(), [
        FlareConversationAction.pin,
        FlareConversationAction.mute,
        FlareConversationAction.markRead,
        FlareConversationAction.archive,
        FlareConversationAction.hide,
        FlareConversationAction.clearHistory,
        FlareConversationAction.delete,
      ]);
      expect(entries.where((e) => e.danger).map((e) => e.action), [
        FlareConversationAction.clearHistory,
        FlareConversationAction.delete,
      ]);
      expect(entries.last.danger, isTrue);
    });
    test('full order with every capability', () {
      expect(ids(base, every), [
        FlareConversationAction.pin,
        FlareConversationAction.mute,
        FlareConversationAction.markUnread,
        FlareConversationAction.archive,
        FlareConversationAction.hide,
        FlareConversationAction.clearHistory,
        FlareConversationAction.delete,
      ]);
    });
  });

  testWidgets('renders labels, emits action, busy disables', (tester) async {
    final received = <String>[];
    Widget build({bool busy = false}) => MaterialApp(
      home: Scaffold(
        body: FlareConversationActionSheet(
          conversation: const FlareConversationActionSnapshot(
            id: 'c1',
            title: '设计评审',
            pinned: true,
            unreadCount: 1,
          ),
          capabilities: all,
          busy: busy,
          onAction: (id, a) => received.add('$id/${a.name}'),
        ),
      ),
    );
    await tester.pumpWidget(build());
    expect(find.text('取消置顶'), findsOneWidget);
    expect(find.text('标为已读'), findsOneWidget);
    expect(find.text('删除'), findsOneWidget);
    expect(find.text('置顶'), findsNothing);

    await tester.tap(find.text('删除'));
    await tester.pump();
    expect(received, ['c1/delete']);

    await tester.pumpWidget(build(busy: true));
    await tester.tap(find.text('取消置顶'));
    await tester.pump();
    expect(received, ['c1/delete']);
    expect(tester.takeException(), isNull);
  });

  testWidgets(
    'mark-unread and clear-history read the strings table and emit their ids',
    (tester) async {
      final received = <String>[];
      Widget build(FlareStrings strings) => FlareStringsScope(
        strings: strings,
        child: MaterialApp(
          home: Scaffold(
            body: FlareConversationActionSheet(
              conversation: base,
              capabilities: every,
              onAction: (id, a) => received.add('$id/${a.name}'),
            ),
          ),
        ),
      );
      await tester.pumpWidget(build(const FlareStrings()));
      expect(find.text('标为未读'), findsOneWidget);
      expect(find.text('清空本地记录'), findsOneWidget);
      expect(find.text('标为已读'), findsNothing);
      // The danger group is drawn after the primary group, clear history first.
      expect(
        tester.getTopLeft(find.text('清空本地记录')).dy,
        lessThan(tester.getTopLeft(find.text('删除')).dy),
      );
      expect(
        tester.getTopLeft(find.text('隐藏')).dy,
        lessThan(tester.getTopLeft(find.text('清空本地记录')).dy),
      );

      await tester.tap(find.text('标为未读'));
      await tester.tap(find.text('清空本地记录'));
      await tester.pump();
      expect(received, ['c1/markUnread', 'c1/clearHistory']);

      await tester.pumpWidget(
        build(
          const FlareStrings().copyWith(
            conversationActionSheetMarkUnread: 'Mark as unread',
            conversationActionSheetClearHistory: 'Clear local history',
          ),
        ),
      );
      expect(find.text('Mark as unread'), findsOneWidget);
      expect(find.text('Clear local history'), findsOneWidget);
    },
  );

  testWidgets('empty capabilities show the empty text, not a blank menu', (
    tester,
  ) async {
    await tester.pumpWidget(
      const MaterialApp(
        home: Scaffold(body: FlareConversationActionSheet(conversation: base)),
      ),
    );
    expect(find.text('暂无可用操作'), findsOneWidget);
  });
}

File: packages/android-im-ui/src/test/kotlin/com/flare/im/ui/ConversationActionSheetTest.kt
package com.flare.im.ui

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class ConversationActionSheetTest {
    private val base = FlareConversationActionSnapshot("c1", "设计评审")
    private val all = FlareConversationActionCapabilities(pin = true, mute = true, markRead = true, archive = true, delete = true, hide = true)
    private val every = all.copy(markUnread = true, clearHistory = true)
    private fun ids(c: FlareConversationActionSnapshot, caps: FlareConversationActionCapabilities) =
        conversationActions(c, caps).map { it.action }

    @Test fun nothingWithoutCapabilities() {
        assertEquals(emptyList(), ids(base, FlareConversationActionCapabilities()))
    }

    @Test fun eachCapabilityRevealsItsAction() {
        assertEquals(listOf(FlareConversationAction.Pin), ids(base, FlareConversationActionCapabilities(pin = true)))
        assertEquals(listOf(FlareConversationAction.Mute), ids(base, FlareConversationActionCapabilities(mute = true)))
        assertEquals(listOf(FlareConversationAction.Archive), ids(base, FlareConversationActionCapabilities(archive = true)))
        assertEquals(listOf(FlareConversationAction.Hide), ids(base, FlareConversationActionCapabilities(hide = true)))
        assertEquals(listOf(FlareConversationAction.ClearHistory), ids(base, FlareConversationActionCapabilities(clearHistory = true)))
        assertEquals(listOf(FlareConversationAction.Delete), ids(base, FlareConversationActionCapabilities(delete = true)))
    }

    @Test fun invertsByState() {
        assertEquals(listOf(FlareConversationAction.Unpin), ids(base.copy(pinned = true), FlareConversationActionCapabilities(pin = true)))
        assertEquals(listOf(FlareConversationAction.Unmute), ids(base.copy(muted = true), FlareConversationActionCapabilities(mute = true)))
        assertEquals(listOf(FlareConversationAction.Unarchive), ids(base.copy(archived = true), FlareConversationActionCapabilities(archive = true)))
    }

    @Test fun markReadOnlyWithUnread() {
        assertEquals(emptyList(), ids(base, FlareConversationActionCapabilities(markRead = true)))
        assertEquals(listOf(FlareConversationAction.MarkRead), ids(base.copy(unreadCount = 3), FlareConversationActionCapabilities(markRead = true)))
    }

    @Test fun markUnreadOnlyWhenNothingIsUnread() {
        val both = FlareConversationActionCapabilities(markRead = true, markUnread = true)
        assertEquals(listOf(FlareConversationAction.MarkUnread), ids(base, both))
        assertEquals(listOf(FlareConversationAction.MarkRead), ids(base.copy(unreadCount = 3), both))
        assertEquals(emptyList(), ids(base.copy(unreadCount = 3), FlareConversationActionCapabilities(markUnread = true)))
    }

    @Test fun orderAndDangerGroup() {
        val entries = conversationActions(base.copy(unreadCount = 2), all.copy(clearHistory = true))
        assertEquals(
            listOf(FlareConversationAction.Pin, FlareConversationAction.Mute, FlareConversationAction.MarkRead,
                FlareConversationAction.Archive, FlareConversationAction.Hide, FlareConversationAction.ClearHistory,
                FlareConversationAction.Delete),
            entries.map { it.action },
        )
        assertEquals(listOf(FlareConversationAction.ClearHistory, FlareConversationAction.Delete), entries.filter { it.danger }.map { it.action })
        assertTrue(entries.last().danger)
    }

    @Test fun fullOrderWithEveryCapability() {
        assertEquals(
            listOf(FlareConversationAction.Pin, FlareConversationAction.Mute, FlareConversationAction.MarkUnread,
                FlareConversationAction.Archive, FlareConversationAction.Hide, FlareConversationAction.ClearHistory,
                FlareConversationAction.Delete),
            ids(base, every),
        )
    }

    @Test fun markUnreadAndClearHistoryCopyComesFromTheStringsTable() {
        val defaults = FlareStrings()
        assertEquals("标为未读", defaults.conversationActionSheetMarkUnread)
        assertEquals("清空本地记录", defaults.conversationActionSheetClearHistory)
        val english = FlareStrings {
            conversationActionSheetMarkUnread = "Mark as unread"
            conversationActionSheetClearHistory = "Clear local history"
        }
        assertEquals("Mark as unread", english.conversationActionSheetMarkUnread)
        assertEquals("Clear local history", english.conversationActionSheetClearHistory)
    }

    @Test fun eachActionDrawsItsOwnRegistryGlyph() {
        assertEquals(
            listOf("pin", "unpin", "mute", "notification", "read", "mark-unread", "archive", "unarchive", "eye-off", "clear-history", "delete"),
            FlareConversationAction.entries.map(::conversationActionIconName),
        )
        for (action in FlareConversationAction.entries) {
            assertTrue(conversationActionIconName(action) in flareIconNames, "$action draws a registry glyph")
        }
        // Clearing local history must not look like deleting the conversation.
        assertTrue(flareIconVector(conversationActionIconName(FlareConversationAction.ClearHistory)) != flareIconVector("delete"))
    }
}
